design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,319 @@
+
+### Non-Goals
+
+* Remove the `CertificateRequest` resource before `CertificateSigningRequest` is
+    fully supported and a migration path is in place
+
+---
+
+## Reasons for Approach
+
+Below are a summary of properties of the `CertificateSigningRequest` which
+cause issues or caveats which need to be addressed when adding support for the
+resource.
+
+### Read all access to CertificateSigningRequests
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. This means that in order for users to have the
+same debugging and usage capabilities of the `CertificateRequest`, users will
+have the ability to list all requests in the cluster, rather than being scoped to
+specific namespaces. This is a regression in security.
+
+A solution to this problem could be given to users by providing examples or
+guides to external tooling (e.g. [Open Policy Agent
+(OPA)](https://github.com/open-policy-agent/opa), whereby cluster administrators
+can limit users access to certain `CertificateSigningRequests` based on label
+selectors.
+
+
+### Namespaced issuers referenced by CertificateSigningRequest
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. cert-manager has the concept of namespaced
+issuers that may only be referenced by `CertificateRequests` that reside in
+the same namespace. Since `CertificateSigningRequests` can reference any
+namespaced `Issuer`, there must be some mechanism to prevent privilege
+escalation (Alice references an issuer in Bob's namespace).
+
+cert-manager will enforce an
+[RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) noun and
+verb whereby the requester must have this role bound to them, in order for the
+`CertificateSigningRequest` referencing a namespaced `Issuer` be approved by the
+cert-manager controller. See [here](#conditions).
+
+This will be done via a
+[`SubjectAccessReview`](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/authorization/v1/types.go#L52)
+against the `UserInfo` fields on the `CertificateSigningRequest`. The required
+role must be of `certificates.k8s.io/CertificateSigningRequest` with the verb
+`create` within the target namespace.
+
+Since `Certificates` are also namespaced resources, cert-manager validates that
+created `Certificate` resources may only reference `Issuers` in the same
+namespace, or `ClusterIssuers`. Any resulting `CertificateSigningRequests` will
+reference `Issuers` in the same namespace.
+
+### Duration Field
+
+`CertificateSigningRequests` do not include a `duration` field. To have parity
+with the `CertificateRequest` resource, the duration field will be moved to an
+annotation as `cert-manager.io/duration` which includes a [Go time duration
+string](https://golang.org/pkg/time/#Duration.String). This annotation may not
+be changed after creation of the `CertificateSigningRequest`.
```

Comment:
Curious how you planned on enforcing this.

Reply:
This would be enforced in the admission webhook.

Reply:
How successful have you all been with regards to running admissions webhooks?  In particular, have you had issues with dependency cycles?

The case that I am worried about is an admission webhook's deployment running on Kube with Kubelets using the CSR API to do cert rotation.  The webhook could break that rotation, take the Kubelet offline, which could take the webhook's deployment offline, which would eventually take the cluster offline as cert rotation fails.

Reply:
Yeah I think we need to be very careful adding admission to CSR resources - perhaps we (for now) just do not enforce this, and rely on the `durationHint` field in future having this validation in place? (if it changes after issuance, or even during, whatever the signer observes first would be used/non-deterministic behaviour)